api/tasks.go

```diff
@@ -157,36 +230,91 @@ type Task struct {
 	Services        []Service
 	Resources       *Resources
 	Meta            map[string]string
-	KillTimeout     time.Duration
+	KillTimeout     *time.Duration
 	LogConfig       *LogConfig
 	Artifacts       []*TaskArtifact
 	Vault           *Vault
 	Templates       []*Template
 	DispatchPayload *DispatchPayloadConfig
 }
 
+func (t *Task) Canonicalize() {
+	if t.LogConfig == nil {
+		t.LogConfig = DefaultLogConfig()
+	} else {
+		t.LogConfig.Canonicalize()
+	}
+	if t.Vault != nil {
+		t.Vault.Canonicalize()
+	}
+	for _, artifact := range t.Artifacts {
+		artifact.Canonicalize()
+	}
+	for _, tmpl := range t.Templates {
+		tmpl.Canonicalize()
+	}
+
+	if t.KillTimeout == nil {
+		t.KillTimeout = helper.TimeToPtr(5 * time.Second)
+	}
+
+	min := MinResources()
+	min.Merge(t.Resources)
+	t.Resources = min
+}
+
 // TaskArtifact is used to download artifacts before running a task.
 type TaskArtifact struct {
-	GetterSource  string
+	GetterSource  *string
 	GetterOptions map[string]string
-	RelativeDest  string
+	RelativeDest  *string
+}
+
+func (a *TaskArtifact) Canonicalize() {
+	if a.RelativeDest == nil {
+		a.RelativeDest = helper.StringToPtr("local/")
+	}
 }
 
 type Template struct {
-	SourcePath   string
-	DestPath     string
-	EmbeddedTmpl string
-	ChangeMode   string
-	ChangeSignal string
-	Splay        time.Duration
-	Perms        string
+	SourcePath   *string
+	DestPath     *string
+	EmbeddedTmpl *string
+	ChangeMode   *string
+	ChangeSignal *string
+	Splay        *time.Duration
+	Perms        *string
+}
+
+func (tmpl *Template) Canonicalize() {
+	if tmpl.ChangeMode == nil {
+		tmpl.ChangeMode = helper.StringToPtr("restart")
+	}
+	if tmpl.Splay == nil {
+		tmpl.Splay = helper.TimeToPtr(5 * time.Second)
+	}
+	if tmpl.Perms == nil {
+		tmpl.Perms = helper.StringToPtr("0644")
+	}
 }
 
 type Vault struct {
 	Policies     []string
-	Env          bool
-	ChangeMode   string
-	ChangeSignal string
+	Env          *bool
+	ChangeMode   *string
+	ChangeSignal *string
+}
+
+func (v *Vault) Canonicalize() {
+	if v.Env == nil {
+		v.Env = helper.BoolToPtr(true)
+	}
+	if v.ChangeMode == nil {
+		v.ChangeMode = helper.StringToPtr("restart")
+	}
+	if v.ChangeSignal == nil {
+		v.ChangeSignal = helper.StringToPtr("sighup")
```

Comment:
Capitalize the signal both in that helper and put an else if it is set and upper case it